File: mizan-door/frontend_mobile/lib/models/queue_entry.dart
import 'patient.dart';

enum QueueStatus { waiting, inConsultation, completed, cancelled }

QueueStatus queueStatusFromString(String value) {
  switch (value) {
    case 'waiting':
      return QueueStatus.waiting;
    case 'in_consultation':
      return QueueStatus.inConsultation;
    case 'completed':
      return QueueStatus.completed;
    case 'cancelled':
      return QueueStatus.cancelled;
    default:
      throw ArgumentError('Unknown queue status: $value');
  }
}

class QueueEntry {
  final String id;
  final String clinicId;
  final String patientId;
  final int queueNumber;
  final QueueStatus status;
  final bool isUrgent;
  final DateTime joinedAt;
  final Patient patient;

  const QueueEntry({
    required this.id,
    required this.clinicId,
    required this.patientId,
    required this.queueNumber,
    required this.status,
    required this.isUrgent,
    required this.joinedAt,
    required this.patient,
  });

  factory QueueEntry.fromJson(Map<String, dynamic> json) {
    return QueueEntry(
      id: json['id'] as String,
      clinicId: json['clinic_id'] as String,
      patientId: json['patient_id'] as String,
      queueNumber: json['queue_number'] as int,
      status: queueStatusFromString(json['status'] as String),
      isUrgent: json['is_urgent'] as bool,
      joinedAt: DateTime.parse(json['joined_at'] as String),
      patient: Patient.fromJson(json['patient'] as Map<String, dynamic>),
    );
  }
}

/// Payload broadcast over `WS /ws/clinics/{clinicId}`.
class QueueUpdatedMessage {
  final String clinicId;
  final List<QueueEntry> queue;

  const QueueUpdatedMessage({required this.clinicId, required this.queue});

  factory QueueUpdatedMessage.fromJson(Map<String, dynamic> json) {
    return QueueUpdatedMessage(
      clinicId: json['clinic_id'] as String,
      queue: (json['queue'] as List<dynamic>)
          .map((e) => QueueEntry.fromJson(e as Map<String, dynamic>))
          .toList(),
    );
  }
}

File: mizan-door/backend/app/models/queue_entry.py
import enum
import uuid
from datetime import datetime
from typing import TYPE_CHECKING

from sqlalchemy import Boolean, DateTime, ForeignKey, Integer, func
from sqlalchemy import Enum as SAEnum
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base

if TYPE_CHECKING:
    from app.models.clinic import Clinic
    from app.models.patient import Patient


class QueueStatus(str, enum.Enum):
    WAITING = "waiting"
    IN_CONSULTATION = "in_consultation"
    COMPLETED = "completed"
    CANCELLED = "cancelled"


class QueueEntry(Base):
    __tablename__ = "queue_entries"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)

    clinic_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("clinics.id", ondelete="CASCADE"), nullable=False, index=True
    )
    patient_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("patients.id", ondelete="CASCADE"), nullable=False, index=True
    )

    queue_number: Mapped[int] = mapped_column(Integer, nullable=False)
    status: Mapped[QueueStatus] = mapped_column(
        SAEnum(QueueStatus, name="queue_status", values_callable=lambda enum_cls: [e.value for e in enum_cls]),
        nullable=False,
        default=QueueStatus.WAITING,
    )
    is_urgent: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    joined_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())

    clinic: Mapped["Clinic"] = relationship(back_populates="queue_entries")
    patient: Mapped["Patient"] = relationship(back_populates="queue_entries")
